Add alias to_ary due to it being removed in rack update

carmensea opened the pull request:
## Problem
`to_ary` was removed from `Response` In Rack's update from 2.0.8 to 2.1.1. Stitches inherits from this class and expected a `to_ary` method to be aliased to Response's finish method. 

## Solution
Add the alias that was removed from the update into our class `UnauthorizedResponse`

bwebster commented:
The breaking change to `rack` mentioned in the description can be found [here](https://github.com/rack/rack/pull/1453).

jbusser commented:
> The breaking change to rack mentioned in the description can be found here.

Based on the justification for that PR in Rack, are we making the right change to stitches?

bwebster commented:
I'm wondering the same thing @jbusser.  I was going to post a [link to your suggested change](https://stitchfix.slack.com/archives/C92956NEN/p1578859652009300?thread_ts=1578855666.006500&cid=C92956NEN), which uses `#to_a`.

@carmensea I'd love to hear your thoughts on the best way to proceed.

rossta commented:
Thanks for jumping on this fix, @carmensea 

Just to add additional context for this. `Rack::Response#to_ary` allowed this to work, as is typical with Rack middleware.
```ruby
status, headers, body = @app.call(env)
```

If we are simply going to call `#to_a`, why inherit from `Rack::Response` at all? From what I can see, we don't need additional `Rack::Response` functionality, like chunking, buffering, redirects, etc. https://github.com/rack/rack/blob/d2b8f254a3671a19841e4e9284411b2d414f2675/lib/rack/response.rb#L35-L82 

In other words, perhaps we should just construct the array return value ourselves.

rossta commented:
To answer my own question, my example omits content type and length.

carmensea commented:
@TildeWill and I paired on the @rossta's idea for implementing our own UnauthorizedResponse. If all our responses are Rack Responses, we can treat them the same way. This change deviates from Rack Response and could be problematic if Rack changes their response. Open to suggestions on the best way to proceed.

zorab47 commented:
The `Rack::Response` expects to have `#finish` called to create the response Array. We could add calls to `#finish` in the locations the [`UnauthorizedResponse`](https://github.com/stitchfix/stitches/blob/468e0ea11c1406093a5197c10af8155f5589014c/lib/stitches/api_key.rb#L48) is used.

See https://github.com/rack/rack/blob/d2b8f254a3671a19841e4e9284411b2d414f2675/lib/rack/response.rb#L10-L21, like @rossta mentioned above.

```
  # Rack::Response provides a convenient interface to create a Rack
  # response.
  #
  # It allows setting of headers and cookies, and provides useful
  # defaults (an OK response with empty headers and body).
  #
  # You can use Response#write to iteratively generate your response,
  # but note that this is buffered by Rack::Response until you call
  # +finish+.  +finish+ however can take a block inside which calls to
  # +write+ are synchronous with the Rack response.
  #
  # Your application's +call+ should end returning Response#finish.
```

Thanks for taking on this work! ❤️

carmensea commented:
@rossta and I paired on a different implementation that continues to use Rack::Response. We think this addresses all comments. We used the explicit call to #finish and removed the subclass, because it didn't add value with this change.

bwebster commented on lib/stitches/api_key.rb:
```diff
@@ -71,7 +66,7 @@ def rails_app_module
       status = 401
       body = "Unauthorized - #{reason}"
-      header = { "WWW-Authenticate" => "#{@configuration.custom_http_auth_scheme} realm=#{@realm}" }
+      header = { "WWW-Authenticate" => "#{@configuration.custom_http_auth_scheme} realm=#{rails_app_module}" }
```
👏 Nice simplification.
